Reader: Add Discover Channels

kean opened the pull request:
- Adds "channels" to the "Discover" tab (the set of channel will be updated in the future PRs)
- Add convenience `pinEdges` and `pinCenter` AutoLayout APIs to replace the now deprecated previous inflexible ones
- Remove unused `changeLayoutMargins`
- Remove `readerDiscoverEndpoint` FF

<img width="320" alt="Screenshot 2024-10-17 at 5 11 58 PM" src="https://github.com/user-attachments/assets/5ca90015-085c-4c0f-9711-8b22b334a55a">

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Remove unused changeLayoutMargins

kean committed:
Add new pinEdges method

kean committed:
Add pinCenter

kean committed:
Add initial ReaderDiscoverHeaderView implementation

kean committed:
Remove readerDiscoverEndpoint FF

kean committed:
Rename ReaderDiscoverViewController

kean committed:
Move ReaderDiscoverHeaderView

kean committed:
Fix separator insets in ReaderPostCell

kean committed:
Extract ReaderStreamTitleView

kean committed:
Implement selection in ReaderDiscoverHeaderView

kean committed:
Extract ReaderDiscoverStreamViewController to allow switching between streams

kean committed:
Add initial implementation of Latest Posts

kean committed:
Fix how ghost cells are shown

kean committed:
Add support for tags

kean committed:
Rename tags to channels

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/297">25.5</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

crazytonyli commented on Modules/Sources/WordPressUI/Extensions/UIView+AutoLayout.swift:
```diff
@@ -0,0 +1,100 @@
+    /// Pins edges of the view to the edges of the given container. By default,
+    /// pins to the superview.
+    @discardableResult
+    public func pinEdges(
```
Nitpick: update the signature to be `pin(edges: .all, to: view, ...)` instead of `pinEdges(.all, to: view, ...)`?

kean replied:
It won't be clear what it does with a default argument `view.pin(to: view)`. It has to have the default argument because you'll use `.all` 99% of the time.

crazytonyli commented on Modules/Sources/WordPressUI/Extensions/UIView+AutoLayout.swift:
```diff
@@ -0,0 +1,100 @@
+            assertionFailure("view has to be installed in the view hierarchy")
+            return []
+        }
+        translatesAutoresizingMaskIntoConstraints = false
```
Add an assertion (`self.isDescendant(of: container)`) to make sure `self` is a subview of `container`, because this line may not be appropriate, depending on the relationship between `self` and `container`.

```swift
// OK to set `subview.translatesAutoresizingMaskIntoConstraints = false`
// because subview's frame is set by auto-layout.
subview.pinEdges(to: superview)

// Not okay to set `superview.translatesAutoresizingMaskIntoConstraints = false`
// because we don't know if superview's frame is set by auto-layout or not.
superview.pinEdges(to: subview)
```

kean replied:
> make sure self is a subview of container

It's not a strict requirement. You can pin a view to a layout guide or to a sibling view. It should be pretty clear from the method's name what it does. SnapKit, PureLayout, and a couple of other frameworks all pretty much work the same way, so it shouldn't surprise that that's how it works.

crazytonyli replied:
If there is no hierarchical requirement in these two views, we should remove this line, as I mentioned in the example code above:

```swift
// Not okay to set `superview.translatesAutoresizingMaskIntoConstraints = false`
// because we don't know if superview's frame is set by auto-layout or not.
superview.pinEdges(to: subview)
```

crazytonyli replied:
IIRC UIViewController.view is not framed using auto layout. Maybe you can reproduce the issue I mentioned by using `pinEdges(to: subview)` on a `UIViewController.view`.

kean replied:
Yup, you shouldn't pin the edges of the superview to its child, and I added a bit of docs explaining it. It even sounds kind of backwards, doesn't it?

The default signature `view.pinEdges()` enforces the idea that you are pinning it to a superview because that's the only reasonable option when you don't have a target view.

As a user, you want to avoid writing `translatesAutoresizingMaskIntoConstraints`. It's a pretty common mistake to forget it, which I've also accidentally done many times when using the existing `pinSubviewToAllEdges` methods. If you take pretty much every Auto Layout framework, they are designed to ensure you never have to write `translatesAutoresizingMaskIntoConstraints`. The simple rule is that the receiver of the `pin*` methods is the view you want to layout using Auto Layout → it's safe to set `translatesAutoresizingMaskIntoConstraints` to `false`.

crazytonyli replied:
> As a user, you want to avoid writing translatesAutoresizingMaskIntoConstraints. [...] If you take pretty much every Auto Layout framework, they are designed to ensure you never have to write translatesAutoresizingMaskIntoConstraints.

That's true. I get it's a bit annoying that we are forced to write this one long line for pretty much every single view in the app. However, the framework (or convenient method in this case) should only set the property when it's appropriate to do so. As I mentioned above, it's not appropriate to call `translatesAutoresizingMaskIntoConstraints = true`, if `self` is not framed using auto-layout.

Here is a diff to reproduce the issue:

```diff
diff --git a/WordPress/Classes/ViewRelated/Reader/ReaderDiscoverViewController.swift b/WordPress/Classes/ViewRelated/Reader/ReaderDiscoverViewController.swift
index af9eda30d5..ffa871d863 100644
--- a/WordPress/Classes/ViewRelated/Reader/ReaderDiscoverViewController.swift
+++ b/WordPress/Classes/ViewRelated/Reader/ReaderDiscoverViewController.swift
@@ -101,6 +101,11 @@ class ReaderDiscoverViewController: UIViewController, ReaderDiscoverHeaderViewDe
         view.addSubview(streamVC.view)
         streamVC.view.pinEdges()
         streamVC.didMove(toParent: self)
+
+        let testView = UIView()
+        testView.backgroundColor = .red
+        view.insertSubview(testView, at: 0)
+        view.pinEdges(to: testView) // Reader goes blank.
     }
 
     /// TODO: (tech-debt) the app currently stores the responses from the `/discover`
```

kean replied:
> Here is a diff to reproduce the issue:

I added a runtime check to catch this early:

```swift
#if DEBUG
        if let target = target as? UIView {
            assert(!target.isDescendant(of: self), "The target view can't be a descendant for the view")
        }
#endif
```

crazytonyli commented on Modules/Sources/WordPressUI/Extensions/UIView+AutoLayout.swift:
```diff
@@ -0,0 +1,100 @@
+    @discardableResult
+    public func pinEdges(
+        _ edges: Edge.Set = .all,
+        to container: AutoLayoutItem? = nil,
```
It might be good to make this `container` part of function signature, so that the relationship is clearer: `view.pinEdges(toContainer: anotherView)` instead of `view.pinEdges(to: anotherView)`.

kean replied:
The "container" isn't the best term to use because you can pin the edges to a layout guide `view.pinEdges(to: view.safeAreaLayoutGuide)` or to a sibling view (not a container). I also replaced the variable name it a more generic term "target".

crazytonyli commented on Modules/Sources/WordPressUI/Extensions/UIView+AutoLayout.swift:
```diff
@@ -0,0 +1,100 @@
+            pin(.top, topAnchor.constraint(greaterThanOrEqualTo: container.topAnchor, constant: insets.top))
+            pin(.trailing, trailingAnchor.constraint(lessThanOrEqualTo: container.trailingAnchor, constant: -insets.right))
+            pin(.bottom, bottomAnchor.constraint(lessThanOrEqualTo: container.bottomAnchor, constant: -insets.bottom))
+            pin(.leading, leadingAnchor.constraint(greaterThanOrEqualTo: container.leadingAnchor, constant: insets.left))
```
It doesn't really make a difference in runtime, but using `NSDirectionalEdgeInsets` instead of `UIEdgeInsets` would make this line looks more consistent: `pin(.leading, ....insets.leading)`.

crazytonyli commented on WordPress/Classes/Utility/ContextManager.swift:
```diff
@@ -176,6 +176,13 @@ public class ContextManager: NSObject, CoreDataStack, CoreDataStackSwift {
+        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = NSFetchRequest(entityName: ReaderCard.entityName())
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+        let coordinator = persistentContainer.persistentStoreCoordinator
+        try coordinator.execute(deleteRequest, with: mainContext)
```
[`NSBatchDeleteRequest` API doc](https://developer.apple.com/documentation/coredata/nsbatchdeleterequest) says we need to manually merge the changes to the context object. Does this API do that for us? [Its API doc](https://developer.apple.com/documentation/coredata/nspersistentstorecoordinator/1468872-execute) is pretty short, and doesn't mention that.

kean replied:
I found an existing method that removes cards in `ReaderCardService` and replaced batch delete with it. There aren't going to be many of these cards, so the performance should be fine.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderDiscoverHeaderView.swift:
```diff
@@ -0,0 +1,162 @@
+        scrollView.addSubview(channelsStackView)
+        scrollView.showsHorizontalScrollIndicator = false
+        scrollView.clipsToBounds = false
+        channelsStackView.pinEdges()
```
Nitpick: it's not obvious what views are involved from this line alone. The default parameter values may save us from typing a few words, but personally, I'd prefer the better readability of having _necessary_ parts in a statement. i.e. `channelsStackView.pinEdges(toContainer: self)`

Alternatively, using `margins` instead of `pinEdges` might be better for conveying the idea that insets are adding between the view and its parent: `channelsStackView.margins()`

kean replied:
We could add a separate `pinEdgesToSuperview` method but I didn't want to duplicate this entire signature.

kean committed:
Integarte WPKit changes with code that loads avatars

kean committed:
Add first posts

kean committed:
Add daily prompts

kean committed:
Remove batchDelete

kean committed:
Add a constant for dailyprompt

kean committed:
Add ManagedObjectsObserver

kean committed:
Observe tags

kean commented:
Updates:
- Added remaining channels: "First Posts", "Latest", "Daily Prompts"
- Added observer for tags (ManagedObjectsObserver)
- Added support for new `site_icon` field in Discover

kean committed:
Rename container to target

kean committed:
Add some docs

kean committed:
Add a link to edit your interests to the Reader

kean committed:
Add runtime check for pinEdges

kean committed:
Add .readerDiscoverChannelSelected

kean committed:
Add readerDiscoverEditInterestsTapped

crazytonyli commented on WordPress/Classes/Services/ReaderCardService.swift:
```diff
@@ -129,12 +129,16 @@ class ReaderCardService {
     func clean() {
-        coreDataStack.performAndSave { context in
-            self.removeAllCards(in: context)
+        ReaderCardService.removeAllCards()
```
Pass `self.coreDataStack` to the function.

crazytonyli commented on WordPress/Classes/ViewRelated/Reader/ReaderDiscoverHeaderView.swift:
```diff
@@ -72,6 +82,19 @@ final class ReaderDiscoverHeaderView: UIView {
+
+    // MARK: UITextViewDelegate
+
+    func textView(_ textView: UITextView, shouldInteractWith URL: URL, in characterRange: NSRange) -> Bool {
```
What do you think about adding some defensive code here to make sure only `/interest` URL is handled? Also, maybe add an assertion failure for other URLs.

kean committed:
Update tests

kean committed:
Update unit tests
